data/src/main/java/com/microsoft/azure/kusto/data/ClientDetails.java

```diff
@@ -0,0 +1,166 @@
+
+    /**
+     * Sets the application name and username for Kusto connectors.
+     *
+     * @param name             The name of the connector/application.
+     * @param version          The version of the connector/application.
+     * @param sendUser         True if the user should be sent to Kusto, otherwise "[none]" will be sent.
+     * @param overrideUser     The user to send to Kusto, or null zvto use the current user.
+     * @param appName          The app hosting the connector, or null to use the current process name.
+     * @param appVersion       The version of the app hosting the connector, or null to use "[none]".
+     * @param additionalFields Additional fields to trace.
+     *                         Example: "Kusto.MyConnector:{1.0.0}|App.{connector}:{0.5.3}|Kusto.MyField:{MyValue}"
+     */
+    public static ClientDetails fromConnectorDetails(String name, String version, boolean sendUser, @Nullable String overrideUser, @Nullable String appName,
+            @Nullable String appVersion, Pair<String, String>... additionalFields) {
+        // make an array
+        List<Pair<String, String>> additionalFieldsList = new ArrayList<>();
+        additionalFieldsList.add(Pair.of("Kusto." + name, version));
+
+        if (appName == null) {
+            appName = unpackLazy(defaultApplication);
+        }
+
+        if (appVersion == null) {
+            appVersion = NONE;
+        }
+
+        additionalFieldsList
+                .add(Pair.of(String.format("App.%s", escapeField(appName)), appVersion));
+        if (additionalFields != null) {
+            additionalFieldsList.addAll(Arrays.asList(additionalFields));
+        }
+
+        String app = formatHeader(additionalFieldsList);
+
+        String user = NONE;
+        if (sendUser) {
+            if (overrideUser != null) {
+                user = overrideUser;
+            } else {
+                user = unpackLazy(defaultUser);
+            }
+        }
+
+        return new ClientDetails(app, user, null);
+    }
+
+    private static String getJavaVersion() {
+        String version = System.getProperty("java.version");
+        if (StringUtils.isBlank(version)) {
+            return "UnknownVersion";
+        }
+        return version;
+    }
+
+    private static String getRuntime() {
+        String runtime = System.getProperty("java.runtime.name");
+        if (StringUtils.isBlank(runtime)) {
+            runtime = System.getProperty("java.vm.name");
+        }
+        if (StringUtils.isBlank(runtime)) {
+            runtime = System.getProperty("java.vendor");
+        }
+        if (StringUtils.isBlank(runtime)) {
+            runtime = "UnknownRuntime";
+        }
+
+        return runtime;
+    }
+
+    private String applicationForTracing;
+    private String userNameForTracing;
+    private String appendedClientVersionForTracing;
+
+    public ClientDetails(String applicationForTracing, String userNameForTracing, String appendedClientVersionForTracing) {
```

Comment:
Same for constructor - I'm used to it always being at the top of the class, right after the class' members.